### DISCUSSION TITLE
Which AWS S3 actions are supported by MinIO?

### DISCUSSION THREAD

**Discussion by omarkohl** [minio/minio#16412] [Q&A] [answered]
Is there a resource anywhere that can help me compare which of the AWS actions are supported by MinIO? I was unable to find it.

I am aware of https://min.io/docs/minio/linux/administration/identity-access-management/policy-based-access-control.html#minio-policy-actions but this only lists the S3 actions that may be used in a policy, which is not the same thing as all possible S3 actions.

In AWS, possible S3 actions and S3 actions that may appear in a policy seem to be overlapping sets, but neither is fully contained in the other.

For example s3:BypassGovernanceRetention is a valid policy S3 action key, but is not an S3 action. DeleteBucketIntelligentTieringConfiguration is an S3 action, but can't be use in a policy.

Sources:

* AWS S3 actions: https://docs.aws.amazon.com/AmazonS3/latest/API/API_Operations.html
* AWS S3 Action elements allowed in an IAM policy: https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazons3.html

Some examples where MinIO differs from AWS in the actions that are supported, irrespective of policies:

* DeleteBucket is supported in both
* PutObjectACL is a dummy in MinIO, so it's kind of _partially_ supported
* GetBucketAnalyticsConfiguration is not supported in MinIO (presumably leading to a 404 HTTP error)

If no such detailed comparison exists I suggest creating one. I am also willing to start working on it. Would you be willing to include it in one of the minio Git repositories (e.g. the docs), so other people can benefit from it and keep it up to date?

**Comment by ravindk89** [comment, marked as answer]
The linked page *does* list all (barring a few I might have missed when I put the page together) S3 Policy Actions that MinIO supports - and many link to the corresponding S3 API (what I assume you mean by S3 action). That is probably a starting point for API support.

But in terms of a dedicated "These are all the S3 APIs we have definitive support for" - that doesn't exist _yet_. We historically have expected the majority of users to lean on the SDKs since those handle signature verification properly (something error prone when done by hand). 

Something to actually emulate the functionality of the S3 API docs - something really developer focused with headers, query params, req/res, etc. - would be a massive undertaking for anyone, and I wouldn't recommend going that far.
If this is scoped to a simple list - no OpenAPI, no Swagger, or similar in-depth API reference w/ full req/res bodies and such - then I would suggest creating this as something like `/docs/api/s3-supported-endpoints.md`. A simple table of supported S3 API operations with notes for those where we either break or extend compatibility would suffice, I assume. It's an open repository, and I don't imagine we'd reject this kind of PR - just maybe include a note of when you compiled the list, as I am unsure we can dedicate engineering resources to maintaining it in the long-term.

Hope that gives you a starting point - I have some lofty goals for OpenAPI integrations in the far-flung future, but right now I know we don't have the writer-cycles for it - and I also cannot say that we would have the engineer cycles as well, beyond cursory review.

**Reply by omarkohl** [reply]
Thanks! I'll submit a PR to minio/docs and I'll add a link here once I do, in case someone stumbles upon this discussion.
